### PR TITLE
Remove pip from scripts and use only conda

### PR DESCRIPTION
**Pull request by aseyboldt**
We should not mix conda packages from defaults with conda-forge during testing, and should not use pip with conda.

### CODE REVIEW

**Comment by aseyboldt** [comment]
This is ready for review.
The tests don't pass because it seems I broke travis caching by disabling and enabling it in the PR, but the tests did pass before the last commit that re-enabled caching. This shouldn't be a problem if we merge, although we will have to remote the cache on master through the travis settings so that the miniconda environment is recreated.

File: requirements-dev.txt
Comment (AlexAndorra) [inline requirements-dev.txt:1]: Seems like you removed `black_nbconvert`? It's needed for [NB styling](https://github.com/pymc-devs/pymc3/wiki/PyMC's-Jupyter-Notebook-Style)

**Review by AlexAndorra — CHANGES_REQUESTED** [review]
Looks good, except for the removal of `black_nbconvert` in dev requirements. I suppose you did that because it can't be installed through any other channel than pip? 
It seems to me that we should be able to require packages that are only installable via pip though. Thoughts?

**Comment by aseyboldt** [comment]
Ah that's where the black_nbconvert is coming from. Guess I messed up the grep :-)
Would it be ok if we add this in a second PR again, so that we can avoid making the caching issue even more confusing?

**Comment by AlexAndorra** [comment]
Ha ha, that's fine by me (I don't completely understand this caching issue
TBH).
We will probably need to install black_nbconvert with pip though, as I
don't think there is a conda port.

Le ven. 3 juil. 2020 à 19:18, Adrian Seyboldt <notifications@github.com> a
écrit :

**Comment by aseyboldt** [comment]
Yes, if there is no conda package then I guess we'll have to use pip :-)

**Comment by AlexAndorra** [comment]
@aseyboldt there is a conflict in `requirements.txt` now: contextvars is in your branch whereas master doesn't have it. Can I remove contextvars to resolve the conflict?
And IIUC you're gonna solve the caching issue on Travis after this PR is merged, so I can merge even if tests are red?

**Comment by aseyboldt** [comment]
The conflict is fixed. I removed the cache on master (also to get the second py3.6 fix merged). The tests should pass now I hope.

**Review by AlexAndorra — COMMENTED** [review]
Test are failing because of these three packages:
```
PackagesNotFoundError: The following packages are not available from current channels:
  - nbsphinx[version='>=0.4.2']
  - sphinx-autobuild==0.7.1
  - watermark
```
I think it's expected, as `watermark` (used for NB styling) is only installable by pip and `create_testenv.sh` doesn't look there anymore -- it should probably do that in last resort?

**Comment by aseyboldt** [comment]
Turns out keras doesn't support py37 yet: https://github.com/keras-team/keras/issues/13964
It is used in one example notebook: https://github.com/pymc-devs/pymc3/blob/master/docs/source/notebooks/convolutional_vae_keras_advi.ipynb
I guess if we removed it from the requirements-dev file, notebook rerendering would break for that notebook?

**Comment by AlexAndorra** [comment]
Yes it would, but we don't run NBs automatically -- just on demand before releases or (hopefully) more frequently and gradually now, given that NB is not a format that lends itself easily to batch runs. So, I don't think removing Keras is very problematic TBH, all the more so since this NB had already problems running IIRC (cc @michaelosthege).

File: pymc3/__init__.py
Comment (AlexAndorra) [inline pymc3/__init__.py]: Why was this dropped? This settled a pretty nasty bug on MacOs. Do we handle this otherwise now?

Comment (aseyboldt) [reply pymc3/__init__.py]: Oh, this was supposed to end up in a different PR.
I seriously doubt that this solves any issue though, because the context isn't used anywhere. This will be trivial to fix after https://github.com/pymc-devs/pymc3/pull/3991 though.

Comment (AlexAndorra) [reply pymc3/__init__.py]: I don't have your knowledge on these issues, but if I understood the python doc correctly, getting the context was a safer way to change it than to set it explicitly (quite counterintuitive) and I'm pretty confident that solved the bug I and a bunch of users were having and that manifested on the terminal when running MPL in NBs ([see the PR](https://github.com/pymc-devs/pymc3/pull/3919) for more context).
In any case, if there is a better way to handle this problem now, I think it is indeed better to explicitely drop this bit in a dedicated PR than hidden in this one -- do you agree?
Also, how is #3991 related to MacOS?

Comment (aseyboldt) [reply pymc3/__init__.py]: You are completely right, this doesn't belong in this PR, I took it out again.

The problem is that just asking for the context doesn't do anything. It only changes something, if you don't use the objects from mulitprocessing, but from the context itself. But since it is never used, it should have absolutely no effect. If you look at the source of multiprocessing, you can see that get_context just returns an object that already exists in a dict: https://github.com/python/cpython/blob/master/Lib/multiprocessing/context.py#L187

The idea of those contexts is that you can use those instead of the multiprocessing namespace to get different behaviour. So you can create a new pool by using `ctx.Pool` instead of `multiprocessing.Pool`.

#3991 makes the context configurable, so that we don't use multiprocessing directly but through a context.

Comment (AlexAndorra) [reply pymc3/__init__.py]: > You are completely right, this doesn't belong in this PR, I took it out again.

Great, thanks, gonna review then. If approved, is it ready for merge now? (note that we're [adding py36 again](https://github.com/pymc-devs/pymc3/pull/4003) in the end 😬)

> The problem is that just asking for the context doesn't do anything.

I agree with that actually 😅 I was confused when playing with this in a NB at the time -- since it had an effect on the bug I was referring to (or at least _seemed_ to have an effect), I went with this 🤷‍♂️ But all you're saying makes sense.

> #3991 makes the context configurable, so that we don't use multiprocessing directly but through a context.

I may have questions about that but I'll ask them directly on the PR 😉

File: requirements-dev.txt
Comment (AlexAndorra) [reply requirements-dev.txt:1]: Now that Travis cache issue is resolved, couldn't you add back `black_nbconvert` to the file?

**Review by AlexAndorra — APPROVED** [review]
Looks good now, thanks @aseyboldt ! Waiting for your answers to my last two comments and I'll merge then 😉

**Comment by aseyboldt** [comment]
I think we have to wait with this one for #4003, it will need a rebase after that.

**Comment by AlexAndorra** [comment]
Where are we on this @aseyboldt ? I think you can rebase now, and I'll merge when tests pass

**Comment by twiecki** [comment]
Ping @aseyboldt.

**Pull request merged**
